lestarch: reworked linux serial driver into linux uart driver following byte stream driver model

LeStarch opened the pull request:
…

| | |
|:---|:---|
|**_Originating Project/Creator_**| |
|**_Affected Component_**|  |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Builds Without Errors (y/n)_**|  |
|**_Unit Tests Pass (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

This takes `Drv::LinuxSerialDriver` and reworks it into `LinuxUartDriver`.  This includes the following:

1. Renaming for clarity as this specifically deals with UART ports
2. Switched to using buffer manager allocate and deallocate calls for acquiring memory
3. Switched to using `Drv::ByteStreamDriverModel` for `send` and `recv`

## Rationale

First and foremost the driver was essentially implementing the `Drv::ByteStreamModel` anyway.  Refactoring it to match makes it a drop-in replacement in uplink/downlink without changing much other than order of arguments and names.

Second, managing its own buffer pool was an old pattern that we've moved away from in favor of the buffer manager. This now uses that pattern.

Third, UART is a clearer more descriptive name than Serial

## Future Work

Unit tests.  They are disabled in CMake, but perhaps should be brought back.

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+     the DCD signal line is the space voltage.
+     */
+    fd = ::open(device, O_RDWR | O_NOCTTY);  // | O_NDELAY);
+    // fd = open(device, O_RDWR | O_NONBLOCK | O_SYNC);
```
Remove this dead code.

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+                           UartFlowControl fc,
+                           UartParity parity,
+                           bool block) {
+    // TODO remove printf
```
Is this still a TODO?

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+
+    DEBUG_PRINT("Opening UART device %s\n", device);
+
+    // TODO can't use O_NDELAY and block (it is same as O_NONBLOCK), so removing NDELAY for now
```
Remove the word "TODO"

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+        case BAUD_230K:
+            relayRate = B230400;
+            break;
+#if defined TGT_OS_TYPE_LINUX
```
Is this #if defined necessary since this is the LinuxUartDriver?

LeStarch replied:
Yes, as the LinuxUartDriver works on macOS too, albeit limited to standard baud rates.

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+     fcntl(fd, F_SETFL, FNDELAY); before reading to get the same result.
+     */
+    // wait for at least 1 byte for 1 second
+    // TODO VMIN should be 0 when using VTIME, and then it would give the timeout behavior Tim wants
+    cfg.c_cc[VMIN] = 0;    // TODO back to 0
+    cfg.c_cc[VTIME] = 10;  // 1 sec, TODO back to 10
```
Remove/update these TODOs

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+                           UartParity parity,
+                           bool block) {
+    // TODO remove printf
+
```
Assert that device is not nullptr.

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+    // set input mode (non-canonical, no echo,...)
+    newtio.c_lflag = 0;
+
+    // TODO if parity, then do input parity too
```
Is this still a TODO?

LeStarch replied:
I have no idea, I looked at the code and it looked like that was completed....so I removed the todo.

kevin-f-ortega replied:
Thanks. But I still see the TODO comment.

kevin-f-ortega replied:
Could you remove the TODO comment?

LeStarch replied:
Done

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+    }
+
+    // CS8 = 8 data bits, CLOCAL = Local line, CREAD = Enable Receiver
+    // TODO PARENB for parity bit
```
Is this still a TODO?

LeStarch replied:
Same as above.

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+
+    NATIVE_INT_TYPE stat = ::write(this->m_fd, data, xferSize);
+
+    // TODO no need to delay for writes b/c the write blocks
```
Is this TODO still needed?

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+}
+
+void LinuxUartDriver ::serialReadTaskEntry(void* ptr) {
+    Drv::RecvStatus status = RecvStatus::RECV_ERROR;  // added by m.chase 03.06.2017
```
Assert ptr is not nullptr.

LeStarch replied:
done

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+            // Good read:
+            if (stat > 0) {
+                sizeRead = stat;
+                // TODO remove
```
Still needed?

LeStarch replied:
I removed a lot of code there.  This was don't to prevent returning/breaking when not returning the allocated buffer.

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+
+        // check stat, maybe output event
+        if (stat == -1) {
+            // TODO(mereweth) - check errno
```
Still needed?

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+            // TODO(mereweth) - check errno
+            Fw::LogStringArg _arg = comp->m_device;
+            comp->log_WARNING_HI_DR_ReadError(_arg, stat);
+            // comp->serialRecv_out(0,buff,Drv::SER_OTHER_ERR);
```
Is this deadcode?

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+            //              stime.tv_nsec);
+            buff.setSize(sizeRead);
+            status = RecvStatus::RECV_OK;  // added by m.chase 03.06.2017
+            // comp->serialRecv_out(0,buff,Drv::SER_OK);
```
deadcode?

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,404 @@
+                           UartBaudRate baud,
+                           UartFlowControl fc,
+                           UartParity parity,
+                           bool block) {
```
The `block` parameter is not used in this function. Get rid of it.

kevin-f-ortega reviewed (changes requested):
Nice work, Michael. I wrote a couple of comments. Most of them related to deadcode or TODOs.

LeStarch commented:
@timcanham do you concur w/ @kevin-f-ortega that we should also go through and clean-up dead-code and TODOs that helicopter left in here?

timcanham commented:
Agreed!

kevin-f-ortega commented on Drv/LinuxUartDriver/LinuxUartDriver.cpp:
```diff
@@ -0,0 +1,365 @@
+bool LinuxUartDriver::open(const char* const device,
+                           UartBaudRate baud,
+                           UartFlowControl fc,
+                           UartParity parity) {
```
Should assert here that device is not nullptr.

kevin-f-ortega reviewed (changes requested):
Looks great, thank you ! Just one minor change request regarding a TODO comment.

LeStarch commented:
@kevin-f-ortega it should be done now.

kevin-f-ortega reviewed (approved):
LGTM

timcanham commented on Drv/LinuxUartDriver/LinuxUartDriverCommon.cpp:
```diff
@@ -0,0 +1,24 @@
+/*
```
If we're not bothering to have stubs, this could be combined with `LinuxUartDriver.cpp` to keep the file count lower.

timcanham commented on Drv/LinuxUartDriver/LinuxUartDriverStub.cpp:
```diff
@@ -0,0 +1,41 @@
+// ======================================================================
```
Is this really still needed?

LeStarch commented:
@timcanham I fixed that.  Was that just a quick suggestion, or are you finished with the review?

timcanham reviewed (approved):
Didn't see any major issues. Minor optional changes.

timcanham commented:
@LeStarch: Review is done...

LeStarch commented:
@timcanham I did the following:

1. Added comments on the non-standard ports
2. Removed stubs and merged common
3. Removed the test files (they were never built)

This should be ready to go, and I will merge unless I hear from you

LeStarch commented:
@jsoref in the above failed spelling check "Gnd" failed as a miss-spelling.  I went to go spell it out fully....but the action log nor the comments on the files noted that miss-spelling.  Thus, I did not know how to fix it/trace where it was erring and thus I had to add Gnd to the `expect.txt` file.

It isn't a major issue, just curious if you had any thoughts as to why this misspelling wouldn't be listed in the log.

I personally find this action to be quite useful.  Thanks for getting the next version integrated in!

jsoref commented:
<details><summary>Looks like it was the case change from `gnd` to `Gnd` that yielded the unhelpful messaging. It's on my list of things to improve.</summary>

Basically, you had `gnd` in your codebase so `gnd` was in `expect.txt` and if it saw `Gnd` it tolerated that. Your commit removed the last `gnd`, leaving only `Gnd`. The output tried to explain this (but is pretty bad at explaining it).


The classic example is:

1. If you have `MICROSOFT` in your codebase, you'd add `MICROSOFT` to expect, but still want to flag `Microsoft` or `microsoft`.
2. If you then add `Microsoft` to your codebase, it'd complain, and suggest changing `MICROSOFT` to `Microsoft` in `expect.txt` (and it'd then tolerate both `MICROSOFT` and `Microsoft`, but complain about `microsoft`).
3. If you then add `microsoft` to your codebase, it'd complain, and suggest changing `Microsoft` to `microsoft` in `expect.txt` (and it'd then tolerate each of `MICROSOFT`, `Microsoft`, and `microsoft`).
4. You could then freely add/remove `MICROSOFT` / `Microsoft` to/from the codebase w/o it commenting.
5. But when you remove the last instance of `microsoft`, it'd want to switch you back to `Microsoft` (if present) or `MICROSOFT` (if that's the only other remaining one) to return to the stricter rule.

The goal was to reduce the number of near duplicate entries / make the `expect.txt` file shorter. But the confusion of this feature has not really helped end users. So I'm probably just going to turn it off and treat each case independently.

Fwiw, this feature dates to before check-spelling was a thing at all. It's so ingrained in my mental model which hasn't helped me let go of it.
</details>

LeStarch commented:
## Pre-CI Checks: Must perform before launching CI
- [ ] No obviously malicious code introduced
- [ ] PR is against `devel`. Use `Edit -> change branch` to fix, or close PR.

## Code Review
- [ ] Code follows flight software quality guidelines
- [ ] Changes are of reasonable length for review

## Procedural Checks
- [ ] PR passes CI
- [ ] Accept with `squash-and-merge`

LeStarch commented:
@jsoref that explanation helps a lot.  No I understand what to do when I see this case!
